core/math/math_funcs.cpp

```diff
@@ -64,37 +64,35 @@ uint32_t Math::rand() {
 	return rand_from_seed(&default_seed);
 }
 
-double Math::randf() {
 
-	return (double)rand() / (double)Math::RANDOM_MAX;
-}
-
-
-double Math::round(double p_val) {
+int Math::step_decimals(double p_step) {
+	static const int maxn=9;
+	static const double sd[maxn]={
+		0.9999, // somehow compensate for floating point error
+		0.09999,
+		0.009999,
+		0.0009999,
+		0.00009999,
+		0.000009999,
+		0.0000009999,
+		0.00000009999,
+		0.000000009999
+	};
 
-	if (p_val>=0) {
-		return ::floor(p_val+0.5);
-	} else {
-		p_val=-p_val;
-		return -::floor(p_val+0.5);
+	double as=Math::abs(p_step);
+	for(int i=0;i<maxn;i++) {
+		if (as>=sd[i]) {
+			return i;
+		}
 	}
-}
 
-double Math::dectime(double p_value,double p_amount, double p_step) {
-
-	float sgn = p_value < 0 ? -1.0 : 1.0;
-	float val = absf(p_value);
-	val-=p_amount*p_step;
-	if (val<0.0)
-		val=0.0;
-	return val*sgn;
+	return maxn;
 }
 
-
-int Math::step_decimals(double p_step) {
-
+// TODO: use function templates?
+int Math::step_decimals(float p_step) {
```

Comment:
this function needs to be double and use internal logic with doubles, otherwise lack of precision breaks it